### DISCUSSION TITLE
Trick to fake a hard disk temporarily?

### DISCUSSION THREAD

**Discussion by jschwartzenberg** [dosemu2/dosemu2#2279] [Ideas]
Some installers seem to be rather peculiar on what they want to install to. I'm trying to install the Windows 3.1 SDK (1) and it claims it cannot find a hard disk.

I think the drives based on the host filesystem are typically recognized as network ones. While using a hard disk image can be a way to work around this, maybe it would also be possible to have some command that would temporarily let such drives be identified as hard disks instead? Just for that session and/or until the command is used again to undo this?

(1) https://winworldpc.com/product/windows-sdk-ddk/windows-31

**Comment by stsp** [comment]
There is such a command:
`emudrv -d c:` makes C: visible
as a local drive. Unfortunately
in this case we do not support
write, so its read-only.

It may be a dosemu2 bug if some
program fails to install, and you
need to first check if it actually
works with an image. The simplest
bet is probably just unpack an image
with mtools after install.

**Reply by jschwartzenberg** [reply]
> Unfortunately
> in this case we do not support
> write, so its read-only.

Hehe, that makes it somewhat useless for installers :)
I understand it's not trivial to allow write support in this case.


> It may be a dosemu2 bug if some
> program fails to install, and you
> need to first check if it actually
> works with an image.

I'll check that also for this case. If I remember correctly an image allows those SB16 installers to proceed. Maybe it's the same with the win31 sdk.

**Reply by jschwartzenberg** [reply]
But I should create an image with another tool right? What's up with `mkfatimage16`? When I run it, I get this:
```
Usage:
  mkfatimage [-b bsectfile] [{[-t tracks] [-h heads] | -k Kbytes}]
             [-l volume-label] [-f outfile] [-p ] [ -r ] [file...]
DOSEMU$���1��ؽ|�Ѝf��������������"|��؎�1��������ut�����}r���no active partition found
�����read error while reading drive
����partition signature != 55AA
���r�&�>�}U�uω��|��U�A�r2��U�u,��t'�|���|�E
��|�B��|�ø�|�M�u��1۴�^�V<u��#   U��<�IBM  3.3   ��)xV4           FAT12   ��P�[���1۬����������

         the Linux DOS Emulator, 4/1/93
             Version 0.49

         Robert Sanders
         gt8134b@prism.gatech.edu

    See README.first for instructions

    press any key to return to Linux...
U�������
```

I guess I'll have to locate `README.first`!

It seems like a dosemu1 relic that's still lingering in dosemu2 :)

**Reply by stsp** [reply]
> What's up with mkfatimage16?

No idea, ask Andrew, he was developing
it for the last years. :)
Anyway, I'll apply the fix for this garbage
output. Just can't push it yet, as I am trying
to migrate the whole stack to stub v5, and
LP is not doing its job, as usual.

**Reply by jschwartzenberg** [reply]
I found an easier workaround for the SDK rather than using an image file. (The `$_hdimage` syntax is a bit tricky to be honest.)

Running `emudrv -d e:` allows the installer to run and it will still write to `c:\` :) :)

**Reply by stsp** [reply]
Yeah, I remember some progs
wanting at least 1 HD to exist,
even if you don't install to it. I
even wanted to propose that,
but decided it might be too
unlikely to help. But it could. :)

**Reply by jschwartzenberg** [reply]
Yeah, that's a good trick to keep in mind :)

It seems I can't build things using Microsoft's tools within DOSEMU2 btw. Running `nmake` also within a Windows 3.1 hosted dos prompt gives me an error that something requires Windows to run. I guess a lot of this relies on enhanced mode to fully handle dos applications.

I was looking at building these: https://github.com/yeokm1/programming-win31

I'll try it with the Borland compiler the book proposes as well.

**Reply by stsp** [reply]
No, you simply can't run win31
apps from the DOS prompt IIRC.
So its likely you need to run them
somehow differently. In general,
DOSish nmake should not run
any win31 apps I think.

I am quite poor in windows, but
maybe @stefand or someone else
can tell what windoses had the
special command.com that allowed
to run both DOS and windows apps.

**Comment by stefand** [comment]
I don't know much either, but in my experiments with dosbox I could not launch notepad.exe from a DOS prompt even in enhanced mode.

I think Visual Studio 4 (or whatever the last Win16 version was) does it the other way around though: The GUI program launches dos programs (the compilers) and installs a vxd to help with that job. I don't know exactly what the vxd does, I assume it is a helper to capture the output and send it to visual studio instead of spawning a regular dos window. Without that driver Visual Studio refuses to compile anything.

**Comment by stefand** [comment]
Addendum: Windows 95 added the ability to launch windows programs from a DOS prompt; So you could start a dos prompt (command.com, there's no cmd.exe) and run e.g. notepad. Win95 also added 32 bit console programs, so e.g. running ping.exe is a lot fancier than it looks.

**Reply by stsp** [reply]
Do you know what API that
command.com uses? Maybe
we could add such?

**Reply by stefand** [reply]
no idea no :-(

**Reply by stsp** [reply]
But that's an interesting question
actually. If we can do that, we better
do. So for some later research.

**Reply by stefand** [reply]
Also I don't know what Windows NT (pre win2k) does. I don't have a Winnt <= 4 installation. I guess it has cmd.exe and not just legacy command.com (after all they don't want their shell to run in the CPU emulator on non-x86). I have a 32 bit Win7 installation, but the laptop is defective and I need some hot sunlight to warm up some capacitors before I can turn it on.

**Reply by stsp** [reply]
@Baron-von-Riedesel maybe you know? :)

**Reply by stsp** [reply]
I think even NT 3.51 had cmd.exe
and dosx.exe. But my memory can
be wrong on that.

**Reply by stsp** [reply]
IIRC the DOS support didn't change
much in nt3.51..win2k timeframe, and
on XP they added SB 2.0 emulation.
And that was the last DOS update ever,
IIRC.

**Reply by stuaxo** [reply]
> I think even NT 3.51 had cmd.exe and dosx.exe. But my memory can be wrong on that.

OT: but I wonder if cmd.exe comes from OS2 originally.

**Reply by stsp** [reply]
From what I've read, OS2 born
from windows development, but
not from NT branch. So I would
guess that cmd.exe has nothing
to do with os2, but I never installed
os2. :)

**Reply by jschwartzenberg** [reply]
You can play with NT 4 and OS/2 here: https://nepx.github.io/halfix-demo/

In NT 4, while there is `cmd.exe`, when you run it there will be the MS-DOS icon in the corner of the window. I guess they just didn't have a suitable icon though...

**Reply by jschwartzenberg** [reply]
And OS/2 indeed also has `cmd.exe`!

**Reply by stsp** [reply]
> You can play with NT 4 and OS/2 here: https://nepx.github.io/halfix-demo/

Wow, quite cool.
So everyone can write an emulator
even for browser these days, and
why dosemu2 is still not written...

**Reply by jschwartzenberg** [reply]
> why dosemu2 is still not written...

Because it's easy to get distracted gaming in it whereas those other emulators are boring ;)

**Reply by stuaxo** [reply]
I was playing with the ones in on PCJS earlier, so on very early OS2 (before they added a GUI) there is (OS2) cmd.exe and (DOS) command.com.

You have to hit ctrl-esc to get to the TUI to change tasks or launch them, and you can't launch say BASICA from OS2 cmd.exe, only from DOS command.com
https://www.pcjs.org/software/pcx86/sys/os2/ibm/1.0/

I noticed how even back then it has a start command, like Windows NT and Windows 95, was wondering if it was the same one, but realised - that this one can't start DOS programs and it's built in - not start.exe

Then I tried some other versions of Windows there in the really early ones you can start command.com in windows, then run a windows program (e.g. notepad) - I guess because everything is real mode.

Then by Win386 this stops.

In Windows 95 land, I'd assumed maybe whatever API start.exe calls is the one that can run DOS applications if needed, but wasn't so sure after testing it there - if you `start qbasic` in a DOS window it starts as expected, then something different happens if you click start->run and type qbasic: you first get a little window where you can add parameters before it runs it.

**Reply by stsp** [reply]
I assume start.exe is a win32 app,
so no doubts it can start anything.
But command.com in win95 is
probably command.com from DOS7,
so its not a win32 app. Starting
the win32 app from DOS app is
what's interesting to find out how.
